pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java

```diff
@@ -139,10 +143,40 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
         }
         break;
       case STRING:
-        for (String value : values) {
-          int dictId = dictionary.indexOf(value);
-          if (dictId >= 0) {
-            dictIdSet.add(dictId);
+        if (values.size() > RequestContextUtils.THRESHOLD_OF_LARGE_IN_CLAUSE_VALUES) {
+          if (bloomFilter != null) {
+            for (String value : values) {
+              if (bloomFilter.mightContain(value)) {
+                int dictId = dictionary.indexOf(value);
+                if (dictId >= 0) {
+                  dictIdSet.add(dictId);
+                }
+              }
+            }
+          } else {
+            int valueIdx = 0;
+            int dictIdx = 0;
+            while (valueIdx < values.size() && dictIdx < dictionary.length()) {
+              String value = values.get(valueIdx);
+              String dictValue = dictionary.getStringValue(dictIdx);
+              int compareResult = value.compareTo(dictValue);
```

Comment:
Is there a way to make use of the UTF-8 comparison pushdown used in binary searches? It's defined on `ValueReader` - maybe it could be exposed on `Dictionary`? Or `Dictionary` could provide methods to fill a set of dictionary ids given a list of a strings?

Reply:
Thanks! This code block is better. The perf is about 2x than the previous version.